File: src/lib/monday-pg-sync/scripts/sync-pg-to-monday.ts
import mondayApi, { MASTER_PROJECT_COLUMN_MAP, SALES_STAFF_COLUMN_MAP } from '../utils/monday-api';
import db from '../utils/db';
import projectModel from '../models/project';
import customerModel from '../models/customer';
import userModel from '../models/user';

// Constants
const BOARD_ID_USERS = process.env.MONDAY_BOARD_ID_USERS || "5764059860";
const BOARD_ID_DATA = process.env.MONDAY_BOARD_ID_DATA || "6727219152";

// Helper function to log sync operations
async function logSyncOperation(entityType: string, mondayId: string, operation: string, status: string, errorMessage?: string) {
  const query = `
    INSERT INTO sync_logs (entity_type, monday_id, operation, status, error_message)
    VALUES ($1, $2, $3, $4, $5)
  `;
  
  try {
    await db.query(query, [entityType, mondayId, operation, status, errorMessage || null]);
  } catch (error) {
    console.error('Error logging sync operation:', error);
  }
}

// Function to sync users from PostgreSQL to Monday.com
export async function syncUsersToMonday() {
  console.log('🔄 Syncing Users from PostgreSQL to Monday.com...');
  
  try {
    // Get all users from the database
    const users = await userModel.getAllUsers();
    console.log(`Found ${users.length} users to sync to Monday.com`);
    
    // Create a reverse mapping of field names to column IDs
    const reverseColumnMap: Record<string, string> = {};
    for (const [columnId, fieldName] of Object.entries(SALES_STAFF_COLUMN_MAP)) {
      reverseColumnMap[fieldName] = columnId;
    }
    
    // Process each user
    for (const user of users) {
      if (!user.monday_id) {
        console.warn(`Skipping user ${user.id} - no Monday.com ID`);
        continue;
      }
      
      // Prepare column values for Monday.com
      const columnValues: Record<string, string> = {};
      
      // Map user fields to Monday.com columns
      if (user.first_name && reverseColumnMap['first_name']) {
        columnValues[reverseColumnMap['first_name']] = user.first_name;
      }
      
      if (user.last_name && reverseColumnMap['last_name']) {
        columnValues[reverseColumnMap['last_name']] = user.last_name;
      }
      
      if (user.email && reverseColumnMap['email']) {
        columnValues[reverseColumnMap['email']] = user.email;
      }
      
      if (user.phone && reverseColumnMap['phone']) {
        columnValues[reverseColumnMap['phone']] = user.phone;
      }
      
      // Combine address fields for Monday.com
      if (user.address && reverseColumnMap['address']) {
        let fullAddress = user.address;
        if (user.city) fullAddress += `, ${user.city}`;
        if (user.state) fullAddress += `, ${user.state}`;
        if (user.zip) fullAddress += ` ${user.zip}`;
        
        columnValues[reverseColumnMap['address']] = fullAddress;
      }
      
      if (user.shirt_size && reverseColumnMap['shirt_size']) {
        columnValues[reverseColumnMap['shirt_size']] = user.shirt_size;
      }
      
      // Update the user in Monday.com
      try {
        // Update each column individually
        for (const [columnId, value] of Object.entries(columnValues)) {
          await mondayApi.updateMondayItem(BOARD_ID_USERS, user.monday_id, columnId, value);
        }
        
        await logSyncOperation('user', user.monday_id, 'update', 'success');
      } catch (error) {
        console.error(`Error syncing user ${user.id} to Monday.com:`, error);
        await logSyncOperation('user', user.monday_id, 'update', 'failed', (error as Error).message);
      }
    }
    
    console.log('✅ Users sync to Monday.com completed successfully');
    return { success: true, message: 'Users sync to Monday.com completed successfully' };
  } catch (error) {
    console.error('❌ Error syncing Users to Monday.com:', error);
    return { success: false, message: (error as Error).message };
  }
}

// Function to sync projects from PostgreSQL to Monday.com
export async function syncProjectsToMonday() {
  console.log('🔄 Syncing Projects from PostgreSQL to Monday.com...');
  
  try {
    // Get all projects from the database
    const projects = await projectModel.getAllProjects();
    console.log(`Found ${projects.length} projects to sync to Monday.com`);
    
    // Create a reverse mapping of field names to column IDs
    const reverseColumnMap: Record<string, string> = {};
    for (const [columnId, fieldName] of Object.entries(MASTER_PROJECT_COLUMN_MAP)) {
      reverseColumnMap[fieldName] = columnId;
    }
    
    // Process each project
    for (const project of projects) {
      if (!project.monday_id) {
        console.warn(`Skipping project ${project.id} - no Monday.com ID`);
        continue;
      }
      
      // Prepare column values for Monday.com
      const columnValues: Record<string, string> = {};
      
      // Map project fields to Monday.com columns
      if (project.job_progress_link && reverseColumnMap['job_progress_link']) {
        columnValues[reverseColumnMap['job_progress_link']] = project.job_progress_link;
      }
      
      if (project.job_progress_name && reverseColumnMap['job_progress_name']) {
        columnValues[reverseColumnMap['job_progress_name']] = project.job_progress_name;
      }
      
      if (project.job_progress_description && reverseColumnMap['job_progress_description']) {
        columnValues[reverseColumnMap['job_progress_description']] = project.job_progress_description;
      }
      
      if (project.company_cam_link && reverseColumnMap['company_cam_link']) {
        columnValues[reverseColumnMap['company_cam_link']] = project.company_cam_link;
      }
      
      if (project.total_price !== null && reverseColumnMap['total_price']) {
        columnValues[reverseColumnMap['total_price']] = project.total_price.toString();
      }
      
      if (project.total_payment !== null && reverseColumnMap['total_payment']) {
        columnValues[reverseColumnMap['total_payment']] = project.total_payment.toString();
      }
      
      // Get the customer for this project
      const customer = await customerModel.getCustomerByMondayId(project.monday_id);
      
      if (customer) {
        // Map customer fields to Monday.com columns
        if (customer.full_name && reverseColumnMap['full_name']) {
          columnValues[reverseColumnMap['full_name']] = customer.full_name;
        }
        
        if (customer.email && reverseColumnMap['email']) {
          columnValues[reverseColumnMap['email']] = customer.email;
        }
        
        if (customer.phone && reverseColumnMap['phone']) {
          columnValues[reverseColumnMap['phone']] = customer.phone;
        }
        
        // Combine address fields for Monday.com
        if (customer.address && reverseColumnMap['address']) {
          let fullAddress = customer.address;
          if (customer.city) fullAddress += `, ${customer.city}`;
          if (customer.state) fullAddress += `, ${customer.state}`;
          if (customer.zip) fullAddress += ` ${customer.zip}`;
          
          columnValues[reverseColumnMap['address']] = fullAddress;
        }
      }
      
      // Update the project in Monday.com
      try {
        // Update each column individually
        for (const [columnId, value] of Object.entries(columnValues)) {
          await mondayApi.updateMondayItem(BOARD_ID_DATA, project.monday_id, columnId, value);
        }
        
        await logSyncOperation('project', project.monday_id, 'update', 'success');
      } catch (error) {
        console.error(`Error syncing project ${project.id} to Monday.com:`, error);
        await logSyncOperation('project', project.monday_id, 'update', 'failed', (error as Error).message);
      }
    }
    
    console.log('✅ Projects sync to Monday.com completed successfully');
    return { success: true, message: 'Projects sync to Monday.com completed successfully' };
  } catch (error) {
    console.error('❌ Error syncing Projects to Monday.com:', error);
    return { success: false, message: (error as Error).message };
  }
}

// Main sync function
export async function syncPostgresToMonday() {
  console.log('🚀 Starting PostgreSQL to Monday.com sync...');
  
  try {
    // Sync users
    const usersResult = await syncUsersToMonday();
    
    // Sync projects
    const projectsResult = await syncProjectsToMonday();
    
    return {
      success: usersResult.success && projectsResult.success,
      users: usersResult,
      projects: projectsResult
    };
  } catch (error) {
    console.error('❌ Error during sync:', error);
    return {
      success: false,
      message: (error as Error).message
    };
  }
}

// If this script is run directly
if (require.main === module) {
  syncPostgresToMonday()
    .then(result => {
      console.log('Sync result:', result);
      process.exit(result.success ? 0 : 1);
    })
    .catch(error => {
      console.error('Unhandled error during sync:', error);
      process.exit(1);
    });
}

export default {
  syncPostgresToMonday,
  syncUsersToMonday,
  syncProjectsToMonday
};

File: sync-scripts/sync_pg_to_monday.py
import os
import sys
import json
from typing import Dict, List, Any
from monday_api import monday_api, MASTER_PROJECT_COLUMN_MAP, SALES_STAFF_COLUMN_MAP
from db_connector import db
from models import Project, Customer, Property, User, Team

# Board IDs from environment variables
BOARD_ID_USERS = os.environ.get('MONDAY_BOARD_ID_USERS', '5764059860')
BOARD_ID_DATA = os.environ.get('MONDAY_BOARD_ID_DATA', '6727219152')

def reverse_map(mapping: Dict[str, str]) -> Dict[str, str]:
    """Reverse a column mapping dictionary."""
    return {v: k for k, v in mapping.items()}

# Reverse mappings for updating Monday.com
REVERSE_PROJECT_MAP = reverse_map(MASTER_PROJECT_COLUMN_MAP)
REVERSE_USER_MAP = reverse_map(SALES_STAFF_COLUMN_MAP)

def format_address(street: str, city: str, state: str, zip_code: str) -> str:
    """Format address components into a single string."""
    components = []
    if street:
        components.append(street)
    
    location_parts = []
    if city:
        location_parts.append(city)
    if state:
        location_parts.append(state)
    
    location = ", ".join(location_parts)
    if location:
        components.append(location)
    
    if zip_code:
        components.append(zip_code)
    
    return ", ".join(components)

def sync_users_to_monday():
    """Sync users from PostgreSQL to Monday.com."""
    print('🔄 Syncing Users from PostgreSQL to Monday.com...')
    
    try:
        # Get all users with sales_staff role that have been updated recently
        query = """
        SELECT u.*, t.name as team_name
        FROM users u
        LEFT JOIN teams t ON u.team_id = t.id
        WHERE u.role = 'sales_staff'
        ORDER BY u.updated_at DESC
        """
        
        users = db.execute_query(query)
        print(f"Found {len(users)} users to sync")
        
        for user in users:
            user_id = user['id']
            
            # Check if we have a Monday.com mapping for this user
            monday_id = db.get_monday_id('user', user_id)
            
            # Prepare column values for Monday.com
            column_values = {}
            
            # Map PostgreSQL fields to Monday.com columns
            for field, value in user.items():
                if field in REVERSE_USER_MAP:
                    column_id = REVERSE_USER_MAP[field]
                    
                    # Skip None values
                    if value is not None:
                        column_values[column_id] = str(value)
            
            # Special handling for address
            address = format_address(
                user.get('address', ''),
                user.get('city', ''),
                user.get('state', ''),
                user.get('zip', '')
            )
            
            if address and 'address' in REVERSE_USER_MAP:
                column_values[REVERSE_USER_MAP['address']] = address
            
            # Set team name if available
            if user.get('team_name') and 'team_name' in REVERSE_USER_MAP:
                column_values[REVERSE_USER_MAP['team_name']] = user['team_name']
            
            try:
                if monday_id:
                    # Update existing Monday.com item
                    for column_id, value in column_values.items():
                        monday_api.update_monday_item(BOARD_ID_USERS, monday_id, column_id, value)
                    
                    db.log_sync_operation('user', monday_id, 'update', 'success')
                    print(f"✅ Updated user in Monday.com: {user.get('first_name')} {user.get('last_name')}")
                else:
                    # Create new Monday.com item
                    item_name = f"{user.get('first_name', '')} {user.get('last_name', '')}".strip()
                    if not item_name:
                        item_name = user.get('email', 'New User')
                    
                    result = monday_api.create_monday_item(BOARD_ID_USERS, item_name, column_values)
                    new_monday_id = result.get('data', {}).get('create_item', {}).get('id')
                    
                    if new_monday_id:
                        db.add_sync_mapping('user', user_id, new_monday_id)
                        db.log_sync_operation('user', new_monday_id, 'create', 'success')
                        print(f"✅ Created user in Monday.com: {item_name}")
                    else:
                        raise Exception("Failed to get new item ID from Monday.com")
            
            except Exception as e:
                status = 'update' if monday_id else 'create'
                db.log_sync_operation('user', monday_id or 'unknown', status, 'error', str(e))
                print(f"❌ Error syncing user {user_id} to Monday.com: {str(e)}")
        
        print("✅ Users sync to Monday.com completed")
        return True
    
    except Exception as e:
        print(f"❌ Error in users sync to Monday.com: {str(e)}")
        return False

def sync_projects_to_monday():
    """Sync projects from PostgreSQL to Monday.com."""
    print('🔄 Syncing Projects from PostgreSQL to Monday.com...')
    
    try:
        # Get all projects with their related customer and property data
        query = """
        SELECT 
            p.id, p.name, p.job_progress_link, p.job_progress_name, p.job_progress_description,
            p.company_cam_link, p.total_price, p.total_payment, p.current_stage,
            c.id as customer_id, c.full_name, c.email, c.phone,
            prop.id as property_id, prop.street, prop.city, prop.state, prop.zip,
            prop.measurements, prop.one_click_codes,
            n.content as customer_note
        FROM projects p
        LEFT JOIN customers c ON p.customer_id = c.id
        LEFT JOIN properties prop ON c.id = prop.customer_id
        LEFT JOIN notes n ON p.id = n.project_id
        ORDER BY p.updated_at DESC
        """
        
        projects = db.execute_query(query)
        print(f"Found {len(projects)} projects to sync")
        
        for project in projects:
            project_id = project['id']
            
            # Check if we have a Monday.com mapping for this project
            monday_id = db.get_monday_id('project', project_id)
            
            # Prepare column values for Monday.com
            column_values = {}
            
            # Map PostgreSQL project fields to Monday.com columns
            for field, value in project.items():
                if field in REVERSE_PROJECT_MAP:
                    column_id = REVERSE_PROJECT_MAP[field]
                    
                    # Skip None values
                    if value is not None:
                        column_values[column_id] = str(value)
            
            # Special handling for customer fields
            if project.get('full_name') and 'full_name' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['full_name']] = project['full_name']
            
            if project.get('email') and 'email' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['email']] = project['email']
            
            if project.get('phone') and 'phone' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['phone']] = project['phone']
            
            # Format full address from components
            address = format_address(
                project.get('street', ''),
                project.get('city', ''),
                project.get('state', ''),
                project.get('zip', '')
            )
            
            if address and 'address' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['address']] = address
            
            # Add property specific fields
            if project.get('measurements') and 'measurements' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['measurements']] = project['measurements']
            
            if project.get('one_click_codes') and 'one_click_codes' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['one_click_codes']] = project['one_click_codes']
            
            # Add customer note
            if project.get('customer_note') and 'customer_note' in REVERSE_PROJECT_MAP:
                column_values[REVERSE_PROJECT_MAP['customer_note']] = project['customer_note']
            
            try:
                if monday_id:
                    # Update existing Monday.com item
                    for column_id, value in column_values.items():
                        monday_api.update_monday_item(BOARD_ID_DATA, monday_id, column_id, value)
                    
                    db.log_sync_operation('project', monday_id, 'update', 'success')
                    print(f"✅ Updated project in Monday.com: {project.get('name')}")
                else:
                    # Create new Monday.com item
                    item_name = project.get('name') or f"Project for {project.get('full_name', 'Unknown Customer')}"
                    
                    result = monday_api.create_monday_item(BOARD_ID_DATA, item_name, column_values)
                    new_monday_id = result.get('data', {}).get('create_item', {}).get('id')
                    
                    if new_monday_id:
                        db.add_sync_mapping('project', project_id, new_monday_id)
                        db.log_sync_operation('project', new_monday_id, 'create', 'success')
                        print(f"✅ Created project in Monday.com: {item_name}")
                    else:
                        raise Exception("Failed to get new item ID from Monday.com")
            
            except Exception as e:
                status = 'update' if monday_id else 'create'
                db.log_sync_operation('project', monday_id or 'unknown', status, 'error', str(e))
                print(f"❌ Error syncing project {project_id} to Monday.com: {str(e)}")
        
        print("✅ Projects sync to Monday.com completed")
        return True
    
    except Exception as e:
        print(f"❌ Error in projects sync to Monday.com: {str(e)}")
        return False

def main():
    print("🚀 Starting PostgreSQL to Monday.com synchronization")
    
    # Ensure database connection
    try:
        db.connect()
        
        # Ensure required tables exist
        db.create_sync_mapping_table()
        db.create_sync_logs_table()
        
        # Run the sync processes
        sync_users_to_monday()
        sync_projects_to_monday()
        
        print("✅ Synchronization completed successfully")
        
    except Exception as e:
        print(f"❌ Error during synchronization: {str(e)}")
        sys.exit(1)
    finally:
        db.disconnect()

if __name__ == "__main__":
    main()
